x/protocolpool/keeper/msg_server.go

```diff
@@ -22,6 +22,44 @@ func NewMsgServerImpl(keeper Keeper) types.MsgServer {
 	return &MsgServer{Keeper: keeper}
 }
 
+func (k MsgServer) ClaimBudget(ctx context.Context, msg *types.MsgClaimBudget) (*types.MsgClaimBudgetResponse, error) {
+	recipient, err := k.Keeper.authKeeper.AddressCodec().StringToBytes(msg.RecipientAddress)
+	if err != nil {
+		return nil, sdkerrors.ErrInvalidAddress.Wrapf("invalid recipient address: %s", err)
+	}
+
+	amount, err := k.claimFunds(ctx, recipient)
+	if err != nil {
+		return nil, err
+	}
+
+	return &types.MsgClaimBudgetResponse{Amount: amount}, nil
+}
+
+func (k MsgServer) SubmitBudgetProposal(ctx context.Context, msg *types.MsgSubmitBudgetProposal) (*types.MsgSubmitBudgetProposalResponse, error) {
+	if err := k.validateAuthority(msg.GetAuthority()); err != nil {
+		return nil, err
+	}
+
+	recipient, err := k.Keeper.authKeeper.AddressCodec().StringToBytes(msg.RecipientAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := k.validateBudgetProposal(ctx, *msg); err != nil {
+		return nil, err
+	}
+
+	budgetProposal := types.NewBudgetProposal(msg.RecipientAddress, *msg.TotalBudget, msg.StartTime, msg.Tranches, msg.Period)
+
+	// set budget proposal in state
+	err = k.BudgetProposal.Set(ctx, recipient, *budgetProposal)
```

Comment:
What would happen if two budgets to the same address are created?

Reply:
It would be updated with the new budget

Reply:
Should we add this disclaimer in the doc? Current behavior makes sense to me, but may be good to precise.